plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -445,24 +468,26 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
       fc::microseconds                _ro_read_window_effective_time_us{ 0 }; // calculated during option initialization
       std::atomic<int64_t>            _ro_all_threads_exec_time_us; // total time spent by all threads executing transactions. use atomic for simplicity and performance
       fc::time_point                  _ro_read_window_start_time;
+      fc::time_point                  _ro_window_deadline; // only modified on app thread
       boost::asio::deadline_timer     _ro_write_window_timer;
       boost::asio::deadline_timer     _ro_read_window_timer;
       fc::microseconds                _ro_max_trx_time_us{ 0 }; // calculated during option initialization
-      ro_trx_queue_t                  _ro_trx_queue;
-      std::atomic<uint32_t>           _ro_num_active_trx_exec_tasks{ 0 };
-      std::vector<std::future<bool>>  _ro_trx_exec_tasks_fut;
+      ro_trx_queue_t                  _ro_exhausted_trx_queue;
```

Comment:
Why the name change? I don't see why `exhausted`?

Reply:
Because now it only contains previously exhausted read-only transactions.

Reply:
What's an `exhausted read-only transactions`? One where `start_recover_keys` took so long it didn't finish within the read window?
If the read window is very small, couldn't that keep happening?

Reply:
exhausted read-only transaction is one that didn't execute in the window. We require the read-only window to be greater than the max-transaction time, so if executed first in the read-window it should either finish or hit max-transaction-time which does not re-queue.